File: Programmers/Lv4/Lv4_단어퍼즐.cpp
#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <climits>
#include <algorithm>
using namespace std;

// DP 이용
// 맨 뒷자리부터 한글자씩 늘려가며 확인
int solution114(vector<string> strs, string t)
{
	int answer = 0;
	int size = t.length();
	set<string> s;
	vector<int> d(size+1, INT_MAX);

	for (string i : strs)		// set에 단어 조각을 삽입
		s.insert(i);

	d[size] = 0;

	for (int i = size-1; i >= 0; i--) {
		string temp="";
		for (int j = i; j < size; j++) {
			temp += t[j];
			if (s.count(temp) != 0) {		// 있으면
				if(d[j+1]!=INT_MAX)
					d[i] = min(d[i], d[j+1] + 1);
			}
			if (j > i + 5)
				break;
		}
	}
	if (d[0] != INT_MAX)
		answer = d[0];
	else
		answer = -1;

	return answer;
}

File: Programmers/Lv4/Lv4_단어퍼즐.py
# DP
# 留??룹옄由щ????쒓??먯뵫 ?섎젮媛硫??뺤씤
def solution(strs, t):
    answer = 0
    size = len(t)
    s = []
    d = [99999999 for x in range(size+1)]

    for i in strs:      # ?⑥뼱 議곌컖???쎌엯
        s.append(i)

    d[size] = 0
    for i in range(size-1, -1, -1):
        temp = ""
        for j in range(i, size):
            temp += t[j]
            if s.count(temp) != 0:           # ?덉쑝硫?
                if d[j+1] != 99999999:
                    d[i] = min(d[i], d[j+1] + 1)
            if j > i + 5:
                break
    if d[0] != 99999999:
        answer = d[0]
    else:
        answer = -1
    return answer
print(solution(	["ab", "na", "n", "a", "bn"], "nabnabn"))